### ISSUE TITLE
Introspection of pl.Expr for column names

### ISSUE THREAD

**Issue by mgperry** [pola-rs/polars#8471]
### Problem description

Is it possible to programmatically inspect polars expressions from python? I'm writing a library to wrap some polars functionality, and there are some situations where it would be useful to get:

- the name of a column ie `pl.col("foo") -> "foo"`
- check if it's a multiple column select and recover the names (ie `pl.col("foo", "bar") -> ["foo", "bar"]`)
- ditto for typed selections and regex selections
- output name of the column ie `pl.col("foo").[...].alias("bar") -> "bar"`, where possible

To some extent, the first three can be achieved by parsing the string representation but this is not ideal, there will be corner cases and the underlying string conversion could change at any point, I'm guessing.

**Comment by alexander-beedie** [comment]
The first and third of those have first-class support via the `meta` namespace; the second (as yet) does not, though you can at least detect that it's potentially a multi-output expression:
```python
expr = pl.col("foo")
expr.meta.output_name()
# 'foo'

expr = ( (pl.col("foo") * pl.col("bar")) / 2 ).alias("baz")
expr.meta.output_name()
# 'baz'
```
```python
expr = pl.col("^x.*$")
expr.meta.has_multiple_outputs()
# True

expr = pl.col("colx", "coly")
expr.meta.has_multiple_outputs()
# True
```

**Comment by ritchie46** [comment]
>  the second (as yet) does not, though you can at least detect that it's potentially a multi-output expression:

It depends on the schema where it is evaluated. We could add a `materialize` meta function that takes a schema and returns a list of those?

**Comment by alexander-beedie** [comment]
> We could add a `materialize` meta function that takes a schema and returns a list of those?

Makes sense; equivalent to runtime resolution where the schema is available.

**Comment by mgperry** [comment]
Thanks, that's super useful, I'd missed the `meta` namespace in the docs. `materialize` sounds very useful, I guess it could also perform regex / type matching .

I'm curious as to why multiple column (rather than regex or type) selections require schema evaluation?

**Comment by alexander-beedie** [comment]
> I'm curious as to why multiple column (rather than regex or type) selections require schema evaluation?

They probably don't, but currently `output_name()` expects to return a single value, so that case (multiple named columns) is caught-up with the regex and dtype selection overloads for `col` ;)
